### DISCUSSION TITLE
Solution for integrating with proxmox

### DISCUSSION THREAD

**Discussion by swratten** [zitadel/zitadel#5564] [Q&A]
I have configured OpenID Connect with proxmox to use Zitadel as the IDP.
Unfortunately proxmox's openidconnect implementation doesn't seem to support / expose the option to have multiple audiences in the access token response. See below:

pve pvedaemon[432956]: openid authentication failure; rhost=::ffff:192.168.189.227 msg=Failed to verify ID token: Invalid audiences: `207276223515066803` is not a trusted audience

This is because Zitadel adds the project id as an audience for each application / client.

I see where the exception is being logged here:

https://github.com/proxmox/proxmox-openid-rs/blob/bdcecd3214fc11f2a8b96611624c7075eb20a435/src/lib.rs#L219

which is from the openidconnect rust crate's source here (which explains that they explicity throw errors for multiple audiences if not "trusted"):

https://github.com/ramosbugs/openidconnect-rs/blob/64be3c7636bc6f745a9b33f0702950bbacca318c/src/verification.rs#L296

And as far as I can tell through a repo search, the calling code / script is here in proxmox:

https://github.com/proxmox/pve-access-control/blob/fe52bc631c2024fa0d7396830fc99be90905f103/src/PVE/API2/OpenId.pm#L173

I am assuming this is likely something that proxmox needs to workaround or add extra "trusted" audiences to the claims.
This is a potential blocker for my companies adoption of Zitadel and OpenID Connect, so I'm looking to see what my options are or whether there is something already in Zitadel which will get around this issue?

I appreciate this is not really a Zitadel issue, but doubt I will individually convince Proxmox to implement such a feature.

**Comment by fforootd** [comment]
Ha this is a classic issue we see often with integrations.

You could try overwrite the AUD claim with a zitadel action. 
There is an example here https://github.com/zitadel/actions/blob/main/examples/add_claim.js

Let me know if that works

```
// this action can be used in the triggers of flow type `Complement Claims`
function addClaim(ctx, api) {
  // this line adds the claim to the token, if it's not already present
  api.v1.claims('aud', something)
}
```

**Reply by swratten** [reply]
This was my initial thought as well, but what stopped me using it was:
```
// this line adds the claim to the token, if it's not already present
```
I assumed an audience claim would definitely already be present.
I now get in my proxmox syslog:
```
Mar 31 12:04:08 pve pvedaemon[2874432]: openid authentication failure; rhost=::ffff: msg=Failed to contact token endpoint: Failed to parse server response
```

**Reply by swratten** [reply]
I'm fairly sure that the function adds the claim successfully as I get the claim in an id_token for another app that I'm running.
Not sure what proxmox doesn't like about the new response in the code exchange process now.. 
Thanks for the quick reply to the audience issue :)

**Reply by fforootd** [reply]
Hm that's a strange error, do you self-host or do you use the cloud?

**Reply by swratten** [reply]
I self-host v2.19.2, the error is from the proxmox openidconnect lib again, so its now getting even less further than it was from the initial post and not liking the response from the code exchange here. 
https://github.com/proxmox/proxmox-openid-rs/blob/bdcecd3214fc11f2a8b96611624c7075eb20a435/src/lib.rs#L207
The error description doesn't say what part of the server response is the issue so not very helpful.

**Reply by fforootd** [reply]
Ok you could try and enable our access logs https://github.com/zitadel/zitadel/blob/main/cmd/defaults.yaml#L362

With this you could see the http call to the token endpoint being logged in stdout.

This is available in beta since[ 2.20.0](https://github.com/zitadel/zitadel/releases/tag/v2.20.0) so you need to upgrade

**Reply by swratten** [reply]
https://pastebin.com/raw/AjAUWhHD

Some logs, seems to be doing as it suggests... 
Not sure there is much more information to extract apart from that it is getting as far as the proxmox logs suggest.

**Reply by fforootd** [reply]
Hm that is odd, the return form zitadel on /token is a 200, so its hard to tell why proxmox fails 😁

**Reply by swratten** [reply]
Agreed, thanks for the help so far :)

**Reply by swratten** [reply]
Just revisiting this, I am now on Zitadel Version: v2.34.0.
It would appear I can no longer set the 'aud' claim in actions, Is this now expected behavior?
This is in a pre userinfo creation trigger for the compliment token flow.

**Reply by swratten** [reply]
To follow up on this, I added a logger action for the same trigger and flow, and can confirm that when the logger is called, the claims I am setting are logged out. 
But when I inspect the user_id token once delivered to the calling client, the 'aud' is not what I had set it to.

**Reply by lordpayder** [reply]
Same Issue for me, Proxmox is not happy with the audience.
I tried to get it to work with a action by setting the aud = id, but it's not working.
Any tips how to solve this?

**Reply by ibnunowshad** [reply]
I do not see any 'aud' or 'audience' field in Proxmox OIDC Realm

**Comment by isodude** [comment]
Should be resolved by (when resolved): https://github.com/zitadel/zitadel/issues/9200

**Reply by SoFMeRight** [reply]
Should we keep our fingers crossed... 🤣

**Reply by hifabienne** [reply]
Proxmox did a fix on there side, they mentioned, that once the state on the issue is going to patch applied it should work.
You can find more information on this issue here: https://github.com/zitadel/zitadel/issues/8055
We will then also update our docs

**Reply by ibnunowshad** [reply]
I am eagerly waiting for this fix from Proxmox.

**Reply by SoFMeRight** [reply]
They added the audiences field in recent releases now... Still having trouble getting it working. Hopefully someone publishes their success... Simply putting the project ID in the field is not enough for me...

**Reply by hifabienne** [reply]
We've added a comment from a dev at Proxmox, some time ago, to this issue, not yet sure if that still holds true, but it was the guideline on how to do it: https://github.com/zitadel/zitadel/issues/8055#issuecomment-2626798965
